File: app/src/main/java/com/codylimber/fieldphenology/data/generator/DataProcessor.kt
package com.codylimber.fieldphenology.data.generator

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull

data class RawSpecies(
    val scientificName: String,
    val commonName: String,
    val taxonId: Int,
    var totalObs: Int,
    var rarity: String = "",
    var peakWeek: Int = 0,
    var firstWeek: Int = 0,
    var lastWeek: Int = 0,
    var periodCount: Int = 1
)

data class WeeklyRow(
    val species: String,
    val week: Int,
    val n: Int,
    val relAbundance: Float
)

object DataProcessor {

    fun buildSpeciesList(rawCounts: List<JsonObject>): List<RawSpecies> {
        val species = rawCounts.mapNotNull { item ->
            val taxon = item["taxon"]?.jsonObject ?: return@mapNotNull null
            val id = taxon["id"]?.jsonPrimitive?.intOrNull ?: return@mapNotNull null
            RawSpecies(
                scientificName = taxon["name"]?.jsonPrimitive?.contentOrNull ?: "Unknown",
                commonName = taxon["preferred_common_name"]?.jsonPrimitive?.contentOrNull ?: "",
                taxonId = id,
                totalObs = item["count"]?.jsonPrimitive?.intOrNull ?: 0
            )
        }
        computeRarity(species)
        return species
    }

    fun computeRarity(species: List<RawSpecies>) {
        if (species.size <= 3) {
            species.forEach { it.rarity = "Uncommon" }
            return
        }
        val counts = species.map { it.totalObs }.sorted()
        val n = counts.size
        val q25 = counts[n / 4]
        val q75 = counts[3 * n / 4]
        for (sp in species) {
            sp.rarity = when {
                sp.totalObs >= q75 -> "Common"
                sp.totalObs >= q25 -> "Uncommon"
                else -> "Rare"
            }
        }
    }

    fun buildWeeklyMatrix(histograms: Map<String, Map<Int, Int>>): List<WeeklyRow> {
        val rows = mutableListOf<WeeklyRow>()
        for ((speciesName, weekCounts) in histograms) {
            val maxCount = weekCounts.values.maxOrNull()?.coerceAtLeast(1) ?: 1
            for ((week, count) in weekCounts) {
                if (week in 1..53) {
                    rows.add(WeeklyRow(
                        species = speciesName,
                        week = week,
                        n = count,
                        relAbundance = (count.toFloat() / maxCount * 10000).toInt() / 10000f
                    ))
                }
            }
        }
        return rows
    }

    fun detectFlightPeriods(weeklyData: List<WeeklyRow>, speciesName: String, gapWeeks: Int = 3): Int {
        val activeWeeks = weeklyData
            .filter { it.species == speciesName && it.n > 0 }
            .map { it.week }
            .sorted()

        if (activeWeeks.isEmpty()) return 0

        var periods = 1
        for (i in 1 until activeWeeks.size) {
            if (activeWeeks[i] - activeWeeks[i - 1] > gapWeeks) {
                periods++
            }
        }
        return periods
    }
}

File: ios/Manakin/Data/Generator/DataProcessor.swift
import Foundation

class RawSpecies {
    let scientificName: String
    let commonName: String
    let taxonId: Int
    var totalObs: Int
    var rarity: String = ""
    var peakWeek: Int = 0
    var firstWeek: Int = 0
    var lastWeek: Int = 0
    var periodCount: Int = 1

    init(scientificName: String, commonName: String, taxonId: Int, totalObs: Int) {
        self.scientificName = scientificName
        self.commonName = commonName
        self.taxonId = taxonId
        self.totalObs = totalObs
    }
}

struct WeeklyRow {
    let species: String
    let week: Int
    let n: Int
    let relAbundance: Float
}

enum DataProcessor {

    static func buildSpeciesList(rawCounts: [[String: Any]]) -> [RawSpecies] {
        let species = rawCounts.compactMap { item -> RawSpecies? in
            guard let taxon = item["taxon"] as? [String: Any],
                  let id = taxon["id"] as? Int else {
                return nil
            }
            let scientificName = taxon["name"] as? String ?? "Unknown"
            let commonName = taxon["preferred_common_name"] as? String ?? ""
            let totalObs = item["count"] as? Int ?? 0
            return RawSpecies(scientificName: scientificName, commonName: commonName, taxonId: id, totalObs: totalObs)
        }
        computeRarity(species: species)
        return species
    }

    static func computeRarity(species: [RawSpecies]) {
        if species.count <= 3 {
            for sp in species {
                sp.rarity = "Uncommon"
            }
            return
        }
        let counts = species.map { $0.totalObs }.sorted()
        let n = counts.count
        let q25 = counts[n / 4]
        let q75 = counts[3 * n / 4]
        for sp in species {
            if sp.totalObs >= q75 {
                sp.rarity = "Common"
            } else if sp.totalObs >= q25 {
                sp.rarity = "Uncommon"
            } else {
                sp.rarity = "Rare"
            }
        }
    }

    static func buildWeeklyMatrix(histograms: [String: [Int: Int]]) -> [WeeklyRow] {
        var rows: [WeeklyRow] = []
        for (speciesName, weekCounts) in histograms {
            let maxCount = max(weekCounts.values.max() ?? 1, 1)
            for (week, count) in weekCounts {
                if week >= 1 && week <= 53 {
                    let relAbundance = Float(Int(Float(count) / Float(maxCount) * 10000)) / 10000.0
                    rows.append(WeeklyRow(
                        species: speciesName,
                        week: week,
                        n: count,
                        relAbundance: relAbundance
                    ))
                }
            }
        }
        return rows
    }

    static func detectFlightPeriods(weeklyData: [WeeklyRow], speciesName: String, gapWeeks: Int = 3) -> Int {
        let activeWeeks = weeklyData
            .filter { $0.species == speciesName && $0.n > 0 }
            .map { $0.week }
            .sorted()

        if activeWeeks.isEmpty { return 0 }

        var periods = 1
        for i in 1..<activeWeeks.count {
            if activeWeeks[i] - activeWeeks[i - 1] > gapWeeks {
                periods += 1
            }
        }
        // Check for year-wrapping: if first and last active weeks connect across the boundary
        if periods > 1 && activeWeeks.count >= 2 {
            let wrapGap = (53 - activeWeeks.last!) + activeWeeks.first!
            if wrapGap <= gapWeeks {
                periods -= 1 // First and last periods are actually one continuous period
            }
        }
        return periods
    }
}
